<a href="https://colab.research.google.com/github/Wipada6430205202/BSC_DPDM23/blob/main/%E0%B8%AA%E0%B8%B3%E0%B9%80%E0%B8%99%E0%B8%B2%E0%B8%82%E0%B8%AD%E0%B8%87_Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import os

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC_DPDM23'

Mounted at /content/drive


In [ ]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/BSC_DPDM23/reduced_marketbasket.csv'),encoding='latin-1')

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

##แปลง data ให้อยู่ในรูปแบบ Transaction data ด้วย Groupby.apply(list)

In [ ]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [ ]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
data.shape

(541909, 8)

###Quiz หาว่ามีกี่ประเทศ

In [ ]:
data['Country'].value_counts().shape

(38,)

## Association Rules

In [ ]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5954 sha256=520a7aa123fbd99a43ee8f5a18f77930197cd261b45c0d78df46e93b5413b31c
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [ ]:
from apyori import apriori

In [ ]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [ ]:
len(frequentItemsets)

4

In [ ]:
frequentItemsets[0][0]

frozenset({'22423'})

In [ ]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [ ]:
frequentItemsets[0][0].__len__()

1

In [ ]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22697', '22699'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}),

support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])


In [ ]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

In [ ]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['InvoiceNo'])

{'542779',
 '543626',
 '578314',
 '548514',
 '540915',
 '566577',
 '571937',
 '554714',
 '536946',
 '573580',
 '553542',
 '539564',
 '550989',
 '537645',
 'C578144',
 '561631',
 '557262',
 '547938',
 '539436',
 '555743',
 'C562953',
 '579539',
 'C538124',
 '545719',
 '536544',
 '550474',
 '549425',
 '541495',
 '553146',
 '558066',
 '542691',
 '562420',
 '554935',
 '564972',
 '558259',
 '550468',
 '540464',
 '581023',
 '541220',
 '557494',
 '559553',
 '554921',
 '575206',
 '564825',
 '542881',
 '537830',
 '578344',
 '553942',
 '555522',
 '542600',
 '578993',
 '562353',
 'C542413',
 '574076',
 'C537834',
 '575932',
 '570374',
 '574950',
 'C577388',
 '549587',
 '545696',
 '576381',
 '559942',
 '552642',
 'C556268',
 '539708',
 '568567',
 '562955',
 '539635',
 '550213',
 '543201',
 '574298',
 '552514',
 '545543',
 '548163',
 '576246',
 '553035',
 '548212',
 '542349',
 'C545444',
 '553891',
 '559337',
 '561209',
 '579093',
 '545945',
 '577020',
 '538524',
 '541008',
 '550215',
 '554704',
 '

In [ ]:
data[data['InvoiceNo']=='562950']

InvoiceNo StockCode                         Description  Quantity  \
297488    562950     22752        SET 7 BABUSHKA NESTING BOXES         2   
297489    562950     22755     SMALL PURPLE BABUSHKA NOTEBOOK         12   
297490    562950     22753     SMALL YELLOW BABUSHKA NOTEBOOK         12   
297491    562950     22754        SMALL RED BABUSHKA NOTEBOOK         12   
297492    562950     22757        LARGE RED BABUSHKA NOTEBOOK         12   
297493    562950     22758    LARGE PURPLE BABUSHKA NOTEBOOK          12   
297494    562950     22756     LARGE YELLOW BABUSHKA NOTEBOOK         12   
297495    562950     23310             BUBBLEGUM RING ASSORTED        36   
297496    562950     23280  FOLDING BUTTERFLY MIRROR HOT PINK         12   
297497    562950     22383              LUNCH BAG SUKI DESIGN         10   
297498    562950     22492             MINI PAINT SET VINTAGE         36   
297499    562950     71459      HANGING JAM JAR T-LIGHT HOLDER        12   
297500    562950     84945  MULTI COLOUR SILVER T-LIGHT HOLDER        12   
297501    562950     21175         GIN + TONIC DIET METAL SIGN        12   
297502    562950     23164                  REGENCY CAKE SLICE         4   
297503    562950     23245          SET OF 3 REGENCY CAKE TINS         4   
297504    562950     23171            REGENCY TEA PLATE GREEN         12   
297505    562950     22697     GREEN REGENCY TEACUP AND SAUCER         6   

             InvoiceDate  UnitPrice  CustomerID         Country  
297488  08/11/2011 09:56       8.50     15608.0  United Kingdom  
297489  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297490  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297491  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297492  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297493  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297494  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297495  08/11/2011 09:56       0.42     15608.0  United Kingdom  
297496  08/11/2011 09:56       0.83     15608.0  United Kingdom  
297497  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297498  08/11/2011 09:56       0.65     15608.0  United Kingdom  
297499  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297500  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297501  08/11/2011 09:56       2.55     15608.0  United Kingdom  
297502  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297503  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297504  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297505  08/11/2011 09:56       2.95     15608.0  United Kingdom

### Quiz 22386 และ 85099B คือสินค้าอะไร

In [ ]:
set(data[data['StockCode']=='22386']['Description'])

{'JUMBO BAG PINK POLKADOT'}

In [ ]:
set(data[data['StockCode']=='85099B']['Description'])

{'JUMBO BAG RED RETROSPOT'}

#Homework1 หา Frequent Item Set และ Assosiation Rules
-  ของช่วงเวลา Chrismas & New year (1 Dec-31 Dec) [หนามเตย]
-  ของประเทศในทวีป Europe [จจ]
-  ของช่วงเวลา ตรุษจีนในทวีป Asia(15 Jan-15 Feb) [ใบเตย]
-  ของช่วงเวลา Chrismas & New year ในทวีป Europe (1 Dec-31 Dec) [พิม]
-  ของวันวาเลนไทน์ [ตุนนี่]
-  โดยจัดกลุ่มตามประเทศ+เดือน [พิมจ๊อก]

In [ ]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

##จัดการข้อมูล

แยกวัน เดือน ปี เวลา

In [ ]:
import pandas as pd

data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'], format='%m/%d/%Y %H:%M')

# Create new columns for day, month, year, and time
data['Day'] = data['InvoiceDate'].dt.day
data['Month'] = data['InvoiceDate'].dt.month
data['Year'] = data['InvoiceDate'].dt.year
data['Time'] = data['InvoiceDate'].dt.time

# Display the modified DataFrame
print(data.head())

# Drop the original
data.drop(columns=['InvoiceDate'], inplace=True)


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


  InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  Day  Month  \
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom    1     12   
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom    1     12   
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom    1     12   
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom    1     12   
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom    1     12   

   Year      Time  
0  2010  08:26:00  
1  2010  08:26:00  
2  2010  08:26:00  
3  2010  08:26:00  
4  2010  0

##1.ของช่วงเวลา Chrismas & New year (1 Dec-31 Dec)

In [ ]:
import pandas as pd
import os
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

In [ ]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/Mining'

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/Mining/reduced_marketbasket.csv'),encoding='latin-1')

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


FileNotFoundError: [Errno 2] No such file or directory: '/content/drive/MyDrive/Mining/reduced_marketbasket.csv'

In [ ]:
# แปลงคอลัมน์ timestamp เป็น datetime
data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'])

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


KeyError: 'InvoiceDate'

In [ ]:
# แยกวันที่เป็นเดือน, วัน, และปีแยกกัน
data['month'] = data['InvoiceDate'].dt.month
data['day'] = data['InvoiceDate'].dt.day
data['year'] = data['InvoiceDate'].dt.year


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


KeyError: 'InvoiceDate'

In [ ]:
data

In [ ]:
# แปลงข้อมูลเป็น Transaction List
chrismas_and_newyear_transactions= data.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
!pip install apyori
from apyori import apriori

In [ ]:
chrismas_and_newyear_transactions

In [ ]:
min_support = 0.03
christmas_and_newyear_frequentItemsets = list(apriori(chrismas_and_newyear_transactions, min_support=min_support))
for i in christmas_and_newyear_frequentItemsets :
   if i[0].__len__()>1:
    print('Assosiation Rules')
    print(f'frequentItemsets is{i[0]}')
    print(f'have support{i[1]}')
    print(f'othar {i[2:]}')


##2.ของประเทศในทวีป Europe

In [ ]:
import pandas as pd
import os

In [ ]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC_DPDM23'

Mounted at /content/drive


In [ ]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [ ]:
data

In [ ]:
import pandas as pd

data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'], format='%m/%d/%Y %H:%M')

# Create new columns for day, month, year, and time
data['Day'] = data['InvoiceDate'].dt.day
data['Month'] = data['InvoiceDate'].dt.month
data['Year'] = data['InvoiceDate'].dt.year
data['Time'] = data['InvoiceDate'].dt.time

# Display the modified DataFrame
print(data.head())

# Drop the original
data.drop(columns=['InvoiceDate'], inplace=True)

In [ ]:
set(data['Country'])

In [ ]:
EU_country = ['Austria','Belgium','Cyprus','Channel Islands','Czech Republic','Denmark','EIRE','European Community',
              'Finland','France','Germany','Greece','Iceland','Italy','Lithuania','Malta','Netherlands',
              'Norway','Poland','Portugal','Spain','Sweden','Switzerland','United Kingdom']

In [ ]:
EU_data_country = data[data['Country']==""]
for i in range(len(EU_country)):
  x = data[data['Country']==EU_country[i]]
  EU_data_country=pd.concat([x,EU_data_country])

In [ ]:
EU_data_country

In [ ]:
EU_data_country.groupby('InvoiceNo')['StockCode'].apply(list)


In [ ]:
transactions = EU_data_country.groupby('InvoiceNo')['StockCode'].apply(list)


In [ ]:
!pip install apyori

In [ ]:
from apyori import apriori


In [ ]:
apriori(transactions,min_support = 0.053)


In [ ]:
list(apriori(transactions,min_support = 0.053))


In [ ]:
frequentItemsets = list(apriori(transactions,min_support = 0.053))
frequentItemsets

In [ ]:
len(frequentItemsets)


In [ ]:
frequentItemsets[0]


In [ ]:
frequentItemsets[0][0]

In [ ]:
dir(frequentItemsets[0][0])

In [ ]:
frequentItemsets[0][0].__len__()


In [ ]:
min_sup = 0.032

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

In [ ]:
set(EU_data_country[(EU_data_country['StockCode']=='85099B')]['Description'])


In [ ]:
set(EU_data_country[(EU_data_country['StockCode']=='22807')]['Description'])


In [ ]:
transactions =EU_data_country.groupby(['Country', 'Year'])['StockCode'].apply(list)
min_sup = 0.4
frequentItemsets= list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

In [ ]:
set(EU_data_country[(EU_data_country['StockCode']=='22423')]['Year'])


In [ ]:
EU_data_country[EU_data_country['StockCode']=='22252']

In [ ]:
EU_data_country[EU_data_country['InvoiceNo']=='543235']

##3.ของช่วงเวลา ตรุษจีนในทวีป Asia(15 Jan-15 Feb)

In [ ]:
import pandas as pd
import os
!pip install apyori
from apyori import apriori

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC_DPDM23'

In [ ]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/BSC_DPDM23/reduced_marketbasket.csv'),encoding='latin-1')

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
import pandas as pd

data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'], format='%m/%d/%Y %H:%M')

# Create new columns for day, month, year, and time
data['Day'] = data['InvoiceDate'].dt.day
data['Month'] = data['InvoiceDate'].dt.month
data['Year'] = data['InvoiceDate'].dt.year
data['Time'] = data['InvoiceDate'].dt.time


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
Asia = data[(data['Country']=='Bahrain') | (data['Country']=='Hong Kong')| (data['Country']=='Singapore') | (data['Country']=='Japan') | (data['Country']=='Israel') |
            (data['Country']=='Lebanon') | (data['Country']=='Saudi Arabia') | (data['Country']=='United Arab Emirates')]
Asia #แบ่งทวีปเอเชีย


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


InvoiceNo StockCode                     Description  Quantity  \
9783      537218     85016   SET OF 6 VINTAGE NOTELETS KIT         6   
9784      537218     21506      FANCY FONT BIRTHDAY CARD,         12   
9785      537218     22716              CARD CIRCUS PARADE        12   
9786      537218     22983             CARD BILLBOARD FONT        12   
9787      537218     22030          SWALLOWS GREETING CARD        12   
...          ...       ...                             ...       ...   
530276   C580832     21731   RED TOADSTOOL LED NIGHT LIGHT        -2   
530277   C580832     23480  MINI LIGHTS WOODLAND MUSHROOMS       -15   
530278   C580832     22466  FAIRY TALE COTTAGE NIGHT LIGHT       -17   
530279   C580832     23084              RABBIT NIGHT LIGHT        -7   
530280   C580832     22621      TRADITIONAL KNITTING NANCY        -1   

               InvoiceDate  UnitPrice  CustomerID Country  Day  Month  Year  \
9783   2010-12-05 15:46:00       2.55     12763.0   Japan    5     12  2010   
9784   2010-12-05 15:46:00       0.42     12763.0   Japan    5     12  2010   
9785   2010-12-05 15:46:00       0.42     12763.0   Japan    5     12  2010   
9786   2010-12-05 15:46:00       0.42     12763.0   Japan    5     12  2010   
9787   2010-12-05 15:46:00       0.42     12763.0   Japan    5     12  2010   
...                    ...        ...         ...     ...  ...    ...   ...   
530276 2011-12-06 11:40:00       1.45     12753.0   Japan    6     12  2011   
530277 2011-12-06 11:40:00       3.29     12753.0   Japan    6     12  2011   
530278 2011-12-06 11:40:00       1.65     12753.0   Japan    6     12  2011   
530279 2011-12-06 11:40:00       1.79     12753.0   Japan    6     12  2011   
530280 2011-12-06 11:40:00       1.45     12753.0   Japan    6     12  2011   

            Time  
9783    15:46:00  
9784    15:46:00  
9785    15:46:00  
9786    15:46:00  
9787    15:46:00  
...          ...  
530276  11:40:00  
530277  11:40:00  
530278  11:40:00  
530279  11:40:00  
530280  11:40:00  

[1314 rows x 12 columns]

In [ ]:
# สร้าง DataFrame สำหรับเหตุการณ์ Chinese New Year
China = Asia[((Asia['Month'] == 1) & (Asia['Day'] >= 15) & (Asia['Day'] <= 31)) |
            ((Asia['Month'] == 2) & (Asia['Day'] <= 15))]

print(China)

      InvoiceNo StockCode                          Description  Quantity  \
69623    541975     22620          4 TRADITIONAL SPINNING TOPS        60   
69624    541975    85183B  CHARLIE & LOLA WASTEPAPER BIN FLORA         6   
69625    541975    85183A   CHARLIE & LOLA WASTEPAPER BIN BLUE         6   
69626    541975     16236                 KITTY PENCIL ERASERS        28   
69627    541975     21058               PARTY INVITES WOODLAND        12   
...         ...       ...                                  ...       ...   
89595    543911    47590B          PINK HAPPY BIRTHDAY BUNTING        12   
89596    543911     22117     METAL SIGN HER DINNER IS SERVED         12   
89597    543911     82567            AIRLINE LOUNGE,METAL SIGN        12   
89598    543911     21910                  WAY OUT METAL SIGN         12   
89599    543911     21911                   GARDEN METAL SIGN         12   

              InvoiceDate  UnitPrice  CustomerID               Country  Day  \
69623 20

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
trans = China.groupby('InvoiceNo')['Description'].apply(list).reset_index()
print(trans)

  InvoiceNo                                        Description
0    541975  [4 TRADITIONAL SPINNING TOPS, CHARLIE & LOLA W...
1    542102  [GIN & TONIC DIET GREETING CARD , GREEN REGENC...
2    542276  [LAUNDRY 15C METAL SIGN, BEWARE OF THE CAT MET...
3    543179  [MONEY BOX FIRST ADE DESIGN, MONEY BOX KINGS C...
4    543518  [SET OF 3 NOTEBOOKS IN PARCEL, LUNCH BAG DOLLY...
5    543911  [RETROSPOT HEART HOT WATER BOTTLE, DOORMAT UNI...


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
trans_apply = transactions['Description'].apply(lambda x: [str(item) for item in x])
trans_apply

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


KeyError: 'Description'

In [ ]:
min_sup = 0.33 #เปลี่ยนจนกว่าจะเจอ...

frequentItemset_China = list(apriori(trans_apply, min_support = min_sup))
for i in frequentItemset_China:
    if i[0].__len__()>1:
        print('yes')
        print(i)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


NameError: name 'trans_apply' is not defined

In [ ]:
freqItem = list(apriori(trans_apply, min_support=0.33))

for record in freqItem:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()

##4.ของช่วงเวลา Chrismas & New year ในทวีป Europe (1 Dec-31 Dec)

In [ ]:
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules

# Filter data for the Christmas & New Year period and European countries
european_countries = ['Austria','Belgium','Cyprus','Channel Islands','Czech Republic','Denmark','EIRE','European Community',
              'Finland','France','Germany','Greece','Iceland','Italy','Lithuania','Malta','Netherlands',
              'Norway','Poland','Portugal','Spain','Sweden','Switzerland','United Kingdom']

filtered_XmasAndNY = data[(data['Month'] == 12) & (data['Country'].isin(european_countries))]

transaction_XmasAndNY = filtered_XmasAndNY.groupby('InvoiceNo')['StockCode'].apply(list)


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Run Apriori algorithm to find frequent item sets
min_sup=0.03
frequentItemsets = list(apriori(transaction_XmasAndNY,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


AttributeError: 'numpy.dtype[object_]' object has no attribute 'apply'

In [ ]:
set(filtered_XmasAndNY['Country'])

In [ ]:
import pandas as pd

# Given data
relation_record = {
    'items': frozenset({'22086', '22910'}),
    'support': 0.04790221341261976,
    'ordered_statistics': [
        {'items_base': frozenset(), 'items_add': frozenset({'22086', '22910'}),
         'confidence': 0.04790221341261976, 'lift': 1.0},
        {'items_base': frozenset({'22086'}), 'items_add': frozenset({'22910'}),
         'confidence': 0.5160142348754448, 'lift': 7.165023343889778},
        {'items_base': frozenset({'22910'}), 'items_add': frozenset({'22086'}),
         'confidence': 0.6651376146788991, 'lift': 7.165023343889778}
    ]
}

# Create a DataFrame
summary_df = pd.DataFrame(columns=['Product Pair', 'Support', 'Description', 'Confidence', 'Lift'])

# Populate the DataFrame
for stat in relation_record['ordered_statistics']:
    product_pair = relation_record['items']
    support = relation_record['support']
    confidence = stat['confidence']
    lift = stat['lift']

    # Get descriptions based on StockCodes
    descriptions = [data[data['StockCode'] == stock_code]['Description'].iloc[0] for stock_code in product_pair]

    summary_df = summary_df.append({
        'Product Pair': product_pair,
        'Support': support,
        'Description': descriptions,
        'Confidence': confidence,
        'Lift': lift
    }, ignore_index=True)

# Display the summary DataFrame
summary_df

##5.ของวันวาเลนไทน

In [ ]:
!pip install apyori

In [ ]:
from apyori import apriori

In [ ]:
import pandas as pd
import os

In [ ]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC_DPDM23'

In [ ]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/new/reduced_marketbasket.csv.crdownload'),encoding='latin-1')

In [ ]:
data

แยกวัน

In [ ]:
data['InvoiceDate'][0].split()[0]

In [ ]:
date = []
for i in range(len(data['InvoiceDate'])):
  x = data['InvoiceDate'][i].split()[0]
  date.append(x)

In [ ]:
date[0].split('/')

In [ ]:
day =[]
month =[]
year =[]
for i in range(len(data['InvoiceDate'])):
  x = date[i].split('/')[0]
  y = date[i].split('/')[1]
  z = date[i].split('/')[2]
  month.append(x)
  day.append(y)
  year.append(z)

นำ data ใหม่ไปใช้

In [ ]:
data_test_nonclean = data

In [ ]:
data_test_nonclean['date'] = date
data_test_nonclean['day'] = day
data_test_nonclean['month'] = month
data_test_nonclean['year'] = year

In [ ]:
set(data_test_nonclean['date'])

In [ ]:
set(data_test_nonclean['day'])

แปลงวัน

In [ ]:
x=[]
for i in range(1,32):
  x.append(i)
print(x)

In [ ]:
data_test0 = data_test_nonclean.replace(['01', '02', '03', '04', '05', '06', '07', '08', '09', '10','11',
                                         '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22',
                                         '23', '24', '25', '26', '27', '28', '29', '30', '31'],x)

In [ ]:
data_test_nonclean.head()

In [ ]:
data_test0.head()

แปลงเดือน

In [ ]:
set(data_test0['month'])

In [ ]:
data_test = data_test0.replace(to_replace=["1", "2",'3','4','5','6','7','8','9'],value=[1,2,3,4,5,6,7,8,9])

In [ ]:
data_test.isnull().any()

In [ ]:
data_test_nonclean[data_test_nonclean['month']=='01']

In [ ]:
data_test[data_test['month']==1]

In [ ]:
M14_data = data_test[(data_test['month']==2)]

In [ ]:
va_data=M14_data[M14_data['day']==14]

In [ ]:
va_data

จัดกลุ่มตาม Transaction

In [ ]:
transactions_va = va_data.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
len(transactions_va)

In [ ]:
min_sup = 0.45
frequentItemsets_va= list(apriori(transactions_va,min_support = min_sup))
for i in frequentItemsets_va:
    if i[0].__len__()>1:
        print('yes')
        print(i)

จัดกลุ่มตาม Transaction (ประเทศ + เดือน)

In [ ]:
transactions_va

In [ ]:
transactions_va = va_data.groupby(['Country', 'month'])['StockCode'].apply(list)
min_sup = 0.45
frequentItemsets_va= list(apriori(transactions_va,min_support = min_sup))
for i in frequentItemsets_va:
    if i[0].__len__()>1:
        print('yes')
        print(i)

##6.โดยจัดกลุ่มตามประเทศ+เดือน

In [ ]:
transactions_MC = data.groupby(['Country', 'Month'])['StockCode'].apply(list)
transactions_MC

In [ ]:
min_sup = 0.3
frequentItemsets_MC = list(apriori(transactions_MC,min_support = min_sup))
for i in frequentItemsets_MC:
    if i[0].__len__()>1:
        print('yes')
        print(i)

In [ ]:
import pandas as pd

# Given data
relation_records = [
    {
        'items': frozenset({'POST', '22326'}),
        'support': 0.32094594594594594,
        'ordered_statistics': [
            {'items_base': frozenset(), 'items_add': frozenset({'POST', '22326'}),
             'confidence': 0.32094594594594594, 'lift': 1.0},
            {'items_base': frozenset({'22326'}), 'items_add': frozenset({'POST'}),
             'confidence': 0.8407079646017699, 'lift': 1.4638209266007287},
            {'items_base': frozenset({'POST'}), 'items_add': frozenset({'22326'}),
             'confidence': 0.5588235294117647, 'lift': 1.463820926600729}
        ]
    },
    {
        'items': frozenset({'22423', 'POST'}),
        'support': 0.3141891891891892,
        'ordered_statistics': [
            {'items_base': frozenset(), 'items_add': frozenset({'22423', 'POST'}),
             'confidence': 0.3141891891891892, 'lift': 1.0},
            {'items_base': frozenset({'22423'}), 'items_add': frozenset({'POST'}),
             'confidence': 0.6838235294117647, 'lift': 1.1906574394463667},
            {'items_base': frozenset({'POST'}), 'items_add': frozenset({'22423'}),
             'confidence': 0.5470588235294118, 'lift': 1.190657439446367}
        ]
    },
    {
        'items': frozenset({'22630', '22629'}),
        'support': 0.30067567567567566,
        'ordered_statistics': [
            {'items_base': frozenset(), 'items_add': frozenset({'22630', '22629'}),
             'confidence': 0.30067567567567566, 'lift': 1.0},
            {'items_base': frozenset({'22629'}), 'items_add': frozenset({'22630'}),
             'confidence': 0.8811881188118811, 'lift': 2.557173364395263},
            {'items_base': frozenset({'22630'}), 'items_add': frozenset({'22629'}),
             'confidence': 0.8725490196078431, 'lift': 2.557173364395263}
        ]
    }
]

# Create a DataFrame
summary_df = pd.DataFrame(columns=['Product Pair', 'Support', 'Description', 'Confidence', 'Lift'])

# Populate the DataFrame
for record in relation_records:
    product_pair = record['items']
    support = record['support']

    for stat in record['ordered_statistics']:
        confidence = stat['confidence']
        lift = stat['lift']

        # Get descriptions based on StockCodes
        descriptions = [data[data['StockCode'] == stock_code]['Description'].iloc[0] for stock_code in product_pair]

        summary_df = summary_df.append({
            'Product Pair': product_pair,
            'Support': support,
            'Description': descriptions,
            'Confidence': confidence,
            'Lift': lift
        }, ignore_index=True)

# Display the summary DataFrame
summary_df
